### ISSUE TITLE
 Upgrading through RPM deletes important configs

### ISSUE THREAD

**Issue by duritong** [processone/ejabberd#2775]
While Upgrading through the RPM all the necessary (and user-edited) config files are deleted although they are in /opt/ejabberd/conf and thus renders an installation unusable.

This is a re-opening of #2734, as it looks like nobody looks into comments on closed tickets. But there is an easy way to reproduce the issue, and also others are reporting the problem.
Sorry if this is an really an issue on the user side, BUT then it is not obvious how the RPMs should be used and I am showing that I am using it as advised by @cromain  

Please also have a look at the attached Vagrant file to reproduce the issue, that shows how ejabberd.yml is deleted during the upgrade although it is in the directory that @cromain mentioned to be the one for configuration. Meaning everything is in /opt/ejabberd and it's removed during upgrade!

```
$ cat Vagrantfile 
script = <<-SCRIPT
set -x
yum install -y https://www.process-one.net/downloads/ejabberd/18.09/ejabberd-18.09-0.x86_64.rpm
ls -l /opt/ejabberd-18.09/
echo "# Some config" >> /opt/ejabberd/conf/ejabberd.yml
ls -l /opt/ejabberd/conf/ejabberd.yml
yum update -y https://www.process-one.net/downloads/ejabberd/18.12/ejabberd-18.12-0.x86_64.rpm
ls -l /opt/ejabberd-18.12/
ls -l /opt/ejabberd/conf/ejabberd.yml
echo "****"
ls -l /opt/ejabberd/
SCRIPT

Vagrant.configure("2") do |config|
  config.vm.box = "centos/7"
  config.vm.provision "shell", inline: script
end
```

Here is the output of the run

```$ vagrant up
Bringing machine 'default' up with 'virtualbox' provider...
==> default: Importing base box 'centos/7'...
==> default: Matching MAC address for NAT networking...
==> default: Checking if box 'centos/7' is up to date...
==> default: A newer version of the box 'centos/7' for provider 'virtualbox' is
==> default: available! You currently have version '1809.01'. The latest is version
==> default: '1811.02'. Run `vagrant box update` to update.
==> default: Setting the name of the VM: ejabberd-upgrade-problem_default_1545662379496_72773
==> default: Clearing any previously set network interfaces...
==> default: Preparing network interfaces based on configuration...
    default: Adapter 1: nat
==> default: Forwarding ports...
    default: 22 (guest) => 2222 (host) (adapter 1)
==> default: Booting VM...
==> default: Waiting for machine to boot. This may take a few minutes...
    default: SSH address: 127.0.0.1:2222
    default: SSH username: vagrant
    default: SSH auth method: private key
    default: 
    default: Vagrant insecure key detected. Vagrant will automatically replace
    default: this with a newly generated keypair for better security.
    default: 
    default: Inserting generated public key within guest...
    default: Removing insecure key from the guest if it's present...
    default: Key inserted! Disconnecting and reconnecting using new SSH key...
==> default: Machine booted and ready!
==> default: Checking for guest additions in VM...
    default: No guest additions were detected on the base box for this VM! Guest
    default: additions are required for forwarded ports, shared folders, host only
    default: networking, and more. If SSH fails on this machine, please install
    default: the guest additions and repackage the box to continue.
    default: 
    default: This is not an error message; everything may continue to work properly,
    default: in which case you may ignore this message.
==> default: Rsyncing folder: /home/duritong/tmp/ejabberd-upgrade-problem/ => /vagrant
==> default: Running provisioner: shell...
    default: Running: inline script
    default: ++ yum install -y https://www.process-one.net/downloads/ejabberd/18.09/ejabberd-18.09-0.x86_64.rpm
    default: Loaded plugins: fastestmirror
    default: Examining /var/tmp/yum-root-5MZwC6/ejabberd-18.09-0.x86_64.rpm: ejabberd-18.09-0.x86_64
    default: Marking /var/tmp/yum-root-5MZwC6/ejabberd-18.09-0.x86_64.rpm to be installed
    default: Resolving Dependencies
    default: --> Running transaction check
    default: ---> Package ejabberd.x86_64 0:18.09-0 will be installed
    default: --> Finished Dependency Resolution
    default: 
    default: Dependencies Resolved
    default: 
    default: ================================================================================
    default:  Package        Arch         Version       Repository                      Size
    default: ================================================================================
    default: Installing:
    default:  ejabberd       x86_64       18.09-0       /ejabberd-18.09-0.x86_64        27 M
    default: 
    default: Transaction Summary
    default: ================================================================================
    default: Install  1 Package
    default: 
    default: Total size: 27 M
    default: Installed size: 27 M
    default: Downloading packages:
    default: Running transaction check
    default: Running transaction test
    default: Transaction test succeeded
    default: Running transaction
    default:   Installing : ejabberd-18.09-0.x86_64                                      1/1
    default:  
    default:   Verifying  : ejabberd-18.09-0.x86_64                                      1/1
    default:  
    default: 
    default: Installed:
    default:   ejabberd.x86_64 0:18.09-0                                                     
    default: 
    default: Complete!
    default: ++ ls -l /opt/ejabberd-18.09/
    default: total 16
    default: drwxr-xr-x.  2 ejabberd ejabberd 4096 Dec 24 14:40 bin
    default: lrwxrwxrwx.  1 root     root       18 Dec 24 14:40 conf -> /opt/ejabberd/conf
    default: lrwxrwxrwx.  1 root     root       22 Dec 24 14:40 database -> /opt/ejabberd/database
    default: drwxr-xr-x.  3 ejabberd ejabberd 4096 Dec 24 14:40 doc
    default: drwxr-xr-x. 51 ejabberd ejabberd 4096 Dec 24 14:40 lib
    default: drwxr-xr-x.  2 ejabberd ejabberd 4096 Dec 24 14:40 logs
    default: ++ echo '# Some config'
    default: ++ ls -l /opt/ejabberd/conf/ejabberd.yml
    default: -rw-r--r--. 1 ejabberd ejabberd 21897 Dec 24 14:40 /opt/ejabberd/conf/ejabberd.yml
    default: ++ yum update -y https://www.process-one.net/downloads/ejabberd/18.12/ejabberd-18.12-0.x86_64.rpm
    default: Loaded plugins: fastestmirror
    default: Examining /var/tmp/yum-root-5MZwC6/ejabberd-18.12-0.x86_64.rpm: ejabberd-18.12-0.x86_64
    default: Marking /var/tmp/yum-root-5MZwC6/ejabberd-18.12-0.x86_64.rpm as an update to ejabberd-18.09-0.x86_64
    default: Resolving Dependencies
    default: --> Running transaction check
    default: ---> Package ejabberd.x86_64 0:18.09-0 will be updated
    default: ---> Package ejabberd.x86_64 0:18.12-0 will be an update
    default: --> Finished Dependency Resolution
    default: 
    default: Dependencies Resolved
    default: 
    default: ================================================================================
    default:  Package        Arch         Version       Repository                      Size
    default: ================================================================================
    default: Updating:
    default:  ejabberd       x86_64       18.12-0       /ejabberd-18.12-0.x86_64        28 M
    default: 
    default: Transaction Summary
    default: ================================================================================
    default: Upgrade  1 Package
    default: 
    default: Total size: 28 M
    default: Downloading packages:
    default: Running transaction check
    default: Running transaction test
    default: Transaction test succeeded
    default: Running transaction
    default:   Updating   : ejabberd-18.12-0.x86_64                                      1/2
    default:  
    default:   Cleanup    : ejabberd-18.09-0.x86_64                                      2/2
    default:  
    default:   Verifying  : ejabberd-18.12-0.x86_64                                      1/2
    default:  
    default:   Verifying  : ejabberd-18.09-0.x86_64                                      2/2
    default:  
    default: 
    default: Updated:
    default:   ejabberd.x86_64 0:18.12-0                                                     
    default: 
    default: Complete!
    default: ++ ls -l /opt/ejabberd-18.12/
    default: total 16
    default: drwxr-xr-x.  2 ejabberd ejabberd 4096 Dec 24 14:41 bin
    default: lrwxrwxrwx.  1 root     root       18 Dec 24 14:41 conf -> /opt/ejabberd/conf
    default: lrwxrwxrwx.  1 root     root       22 Dec 24 14:41 database -> /opt/ejabberd/database
    default: drwxr-xr-x.  3 ejabberd ejabberd 4096 Dec 24 14:41 doc
    default: drwxr-xr-x. 52 ejabberd ejabberd 4096 Dec 24 14:41 lib
    default: drwxr-xr-x.  2 ejabberd ejabberd 4096 Dec 24 14:41 logs
    default: ++ ls -l /opt/ejabberd/conf/ejabberd.yml
    default: ls: cannot access /opt/ejabberd/conf/ejabberd.yml: No such file or directory
    default: ****
    default: ++ echo '****'
    default: ++ ls -l /opt/ejabberd/
    default: total 12
    default: drwxr-xr-x. 2 ejabberd ejabberd 4096 Dec 24 14:41 conf
    default: drwxr-xr-x. 3 ejabberd ejabberd 4096 Dec 24 14:40 database
    default: drwxr-xr-x. 2 ejabberd ejabberd 4096 Dec 24 14:40 logs

```

Upgrading RPMs should not delete configs and as the output it shows, things are not working as you described when closing the issue.

**Comment by cromain** [comment]
Thanks reopening this. pre-uninstall needs some rework then. This will be fixed by next release.

**Comment by cromain** [comment]
rpm forced removal of conf and database directory. this is now fixed but won't be available after you upgrade 19.02 to 19.x
